My Progress to 1.4

Pop000100 opened the pull request:
It doesn't have the toggle expert mode thing working because Re-Logic set Main.Hardmode to private, and journey mode code resets it every world update tick. Through reflection I have all the other features working though.

Pop000100 committed:
Porting to 1.4

I was wondering if you have any thoughts on the issues I mentioned in the vanilla tweaks thread

Pop000100 committed:
Merge branch 'dev' of https://github.com/Pop000100/BossExpertise into dev

gardenappl commented:
I haven't looked at decompiled Terraria 1.4 yet, so I'm not sure if I can help right now.

gardenappl commented:
Would it help if we disabled the Expert Mode Toggle thing in Journey Mode worlds?

Or, the opposite approach, add another command that toggles Journey Mode?

Pop000100 commented:
We could remove it. That would help If we aren't using journey mode but I think that will require IL editing.
I haven't done IL editing but I probably could learn it.

gardenappl commented:
Let's remove the /expert command for now, maybe add that back in a future update.

Pop000100 committed:
Current Version

- Support Mastermode
- Make clarify that the Accessory slot setting effects both slots
- Change method of overriding the slots since it significantly changed in 1.4

gardenappl commented on BossExpertise.cs:
```diff
@@ -10,22 +10,32 @@ namespace BossExpertise
 	{
 
-		public static bool? FakeExpert = false;
+		public static int CurrentDifficulty = 0;
```
I don't get it. How do we differentiate between fake difficulty for boss AI, and the real difficulty?

Pop000100 replied:
Check it again.

It was a mistake, but it was working because the game resets it for me but I made it explicit.

gardenappl commented on BossExpertise.cs:
```diff
@@ -54,48 +64,83 @@ public override void HandlePacket(BinaryReader reader, int whoAmI)
+				FieldInfo expert = typeof(Main)
 				.GetField("_overrideForExpertMode", BindingFlags.Static | BindingFlags.NonPublic);
-				expert.SetValue(null, givenValue);
+				expert.SetValue(null, true);
```
Maybe better to set master mode to false, just in case

Pop000100 replied:
Done.

Pop000100 committed:
Requested changes + 1

- disable MasterMode for safety
- Push accessory slot setting next to other settings of that type
- Reset Difficulty

gardenappl commented on BossExpertiseSystem.cs:
```diff
@@ -0,0 +1,42 @@
+			if (BossExpertise.CurrentDifficulty > 0) //an extra check just in case
+			{
+				BossExpertise.ActualDifficulty = 0;
+				BossExpertise.HookDifficultyMode(0);
```
So this will always reset the world to Normal mode if the config file says "current faked difficulty is expert  or master"?

Pop000100 replied:
Changed. Made it check the actual difficulty. It wouldn't succeed at changing the world since the variable I am reflecting is temporary at run time so it wouldn't be writing the change to disk.

gardenappl commented on BossExpertise.cs:
```diff
@@ -66,38 +64,92 @@ public override void HandlePacket(BinaryReader reader, int whoAmI)
+				return;
+			}
+			SyncDifficultyMode(difficulty);
+			CurrentDifficulty = difficulty;
```
Here we should change `ActualDifficulty`, not `CurrentDifficulty`, no?

Pop000100 replied:
Yes, it should have been but that is currently never ran because its part of the chat command. Fixed anyway.

gardenappl reviewed (commented):
I'm still a bit confused

Pop000100 committed:
More accurately reset state

Also fix some unused code.
